aten/src/ATen/native/mkldnn/xpu/detail/WoQMatmul.cpp

```diff
@@ -0,0 +1,306 @@
+          "matmul supports [n] or [1] when bias dim is 1 ...");
+      if (b.size(0) == 0) {
+        with_bias = false;
+      } else if (m1.dim() == 3) {
+        b = b.expand({mb, m, n}).contiguous();
+      } else if (m1.dim() == 2) {
+        b = b.expand({1, n}).contiguous();
+      }
+    } else if (b.dim() == 2) {
+      TORCH_CHECK(
+          (b.size(0) == m && b.size(1) == n) ||
+              (b.size(0) == 1 && b.size(1) == n) ||
+              (b.size(0) == m && b.size(1) == 1) ||
+              (b.size(0) == 1 && b.size(1) == 1),
+          "matmul supports [m, n] or [1, n] or [m, 1] or [1, 1] when bias dim is 2 ...");
+      if (b.size(0) == 1 && b.size(1) == 1)
+        b = b.expand({1, n}).contiguous();
+    } else if (b.dim() == 3) {
+      TORCH_CHECK(
+          are_expandable({mb, m, n}, b.sizes()),
+          "matmul bias must be expandable to:",
+          dst.sizes(),
+          " but got:",
+          b.sizes());
+      b = b.expand({mb, m, n}).contiguous();
+    } else if (b.dim() == 0) {
+      TORCH_CHECK(
+          b.numel() == 1, "matmul supports 1 numel when bias dim is [] ...");
+      if (m1.dim() == 3) {
+        b = b.expand({mb, m, n}).contiguous();
+      } else {
+        b = b.expand({1, n}).contiguous();
+      }
+    } else {
+      TORCH_CHECK(0, "unsupported bias dim in matmul ...");
+    }
+  }
+
+  // bias is fused in post-op for quantized path
+  b = b.contiguous(); // avoid reorder 2 times
+
+  // torch tensor of int32 dtype, corresponding to oneDNN s32
+  auto m1_usr_dt = get_onednn_dtype(m1); // half <==> f16
+  //   auto m2_usr_dt = dnnl::memory::data_type::s4; // int32, representing 8xint4
+  auto m2_usr_dt = get_onednn_dtype(m2);
+  auto scale_user_dt = get_onednn_dtype(scale_); // half <==> fp16
+  //   auto zp_user_dt = dnnl::memory::data_type::s4; // int32, representing 8xint4
+  auto zp_user_dt = get_onednn_dtype(zp_);
+  auto dst_usr_dt = get_onednn_dtype(dst); // half <==> f16
+
+  auto m1_dt = m1_usr_dt;
+  auto m2_dt = dnnl::memory::data_type::u4;
+  auto scale_dt = scale_user_dt;
+  auto zp_dt = dnnl::memory::data_type::u4;
+  auto dst_dt = dst_usr_dt;
+  dnnl::memory::data_type bias_dt;
+
+  dnnl::memory::desc m1_md, m1_usr_md;
+  dnnl::memory::desc m2_md, m2_usr_md;
+  dnnl::memory::desc scale_md, scale_usr_md;
+  dnnl::memory::desc zp_md, zp_usr_md;
+  dnnl::memory::desc dst_md, dst_usr_md;
+  dnnl::memory::desc b_md;
+
+  // STEP1: create dnnl::memory desc
+  dnnl::memory::dims m1_dims, m2_dims, m2_usr_dims, scale_dims, zp_dims, zp_usr_dims,
+      dst_dims, bias_dims;
+  dnnl::memory::dims m1_strides, m2_strides, m2_usr_strides, scale_strides,
+      zp_strides, zp_usr_strides, dst_strides, bias_strides;
+
+  const uint64_t num_groups = (uint64_t)(k / group_size);
+  const uint64_t compressed_n = (uint64_t)(n / 8);
+  const uint64_t compressed_k = (uint64_t)(k / 8);
+
+  m2_usr_dims = {compressed_k, n};
+  scale_dims = {num_groups, n};
+  zp_dims = {1};
```

Comment:
The dims of zp_dims is not aligned with the original zp inputs. With this limitation, only the symmetric or per-tensor quantization is supported.   Pls add the comments about this limitation of oneDNN.

Reply:
OneDNN provides us with a way to support asymmetry, allowing us to handle asymmetrical scenarios. I'm currently testing it, and if it works, I will modify it here to support both symmetric and asymmetric logic.